File: src/julia/UniversalAxiom.jl
"""
The Universal Axiom - Core Implementation (Julia)

Intelligence_n = E_n · (1 + F_n) · X · Y · Z · (A · B · C)

Foundation Layer: Impulses (A), Elements (B), Pressure (C)
Dynamic Layer: Exponential Growth (E_n), Fibonacci Sequence (F_n)
Cognitive Layer: Subjectivity Scale (X), Why Axis (Y), TimeSphere (Z)
"""

module UniversalAxiom

export FoundationLayer, DynamicLayer, CognitiveLayer
export Axiom, AxiomSimulator, AxiomState
export compute, compute_intelligence, evolve!, apply_pressure!, adjust_subjectivity!, strengthen_purpose!
export get_state, state_to_dict, simulate_evolution, simulate_contradiction_resolution, get_coherence_metric
export fibonacci_sequence, fibonacci, exponential_growth, objectivity
export ProofStep, ErdosProblem, MathSolutions
export add_proof_step!, add_problem!, get_problem, list_problems, summaries, erdos_seed

"""
Foundation Layer: A · B · C
"""
mutable struct FoundationLayer
    impulses::Float64   # A - Fundamental drives
    elements::Float64   # B - Core components
    pressure::Float64   # C - Constraints and forces
end

function compute(foundation::FoundationLayer)::Float64
    foundation.impulses * foundation.elements * foundation.pressure
end

"""
Dynamic Layer: E_n · (1 + F_n)
"""
mutable struct DynamicLayer
    n::Int64                    # Current iteration/step
    base_exponential::Float64   # Base for exponential growth

    DynamicLayer(n::Int64) = new(n, 3.0)
end

function exponential_growth(dynamic::DynamicLayer)::Float64
    (2 * (dynamic.base_exponential ^ dynamic.n)) - 1
end

function fibonacci(n::Int64)::Int64
    if n <= 1
        return 1
    end

    a, b = 1, 1
    for _ in 2:n
        a, b = b, a + b
    end

    return b
end

function fibonacci(dynamic::DynamicLayer)::Int64
    fibonacci(dynamic.n)
end

function compute(dynamic::DynamicLayer)::Float64
    E_n = exponential_growth(dynamic)
    F_n = fibonacci(dynamic)
    E_n * (1 + F_n)
end

"""
Cognitive Layer: X · Y · Z
"""
mutable struct CognitiveLayer
    subjectivity::Float64  # X - Subjectivity level (0 = objective, 1 = subjective)
    purpose::Float64       # Y - Purpose-driven reasoning strength
    time::Float64          # Z - Temporal continuity and irreversibility
end

function compute(cognitive::CognitiveLayer)::Float64
    # X represents objectivity: (1 - subjectivity)
    objectivity = 1.0 - cognitive.subjectivity
    objectivity * cognitive.purpose * cognitive.time
end

function objectivity(cognitive::CognitiveLayer)::Float64
    1.0 - cognitive.subjectivity
end

"""
Complete state of the Universal Axiom
"""
struct AxiomState
    n::Int64
    foundation::NamedTuple
    dynamic::NamedTuple
    cognitive::NamedTuple
    intelligence::Float64
end

"""
The Universal Axiom - Complete Intelligence Model

Intelligence_n = E_n · (1 + F_n) · X · Y · Z · (A · B · C)
"""
mutable struct Axiom
    foundation::FoundationLayer
    dynamic::DynamicLayer
    cognitive::CognitiveLayer
    n::Int64

    function Axiom(;
        impulses::Float64 = 1.0,
        elements::Float64 = 1.0,
        pressure::Float64 = 1.0,
        subjectivity::Float64 = 0.0,
        purpose::Float64 = 1.0,
        time::Float64 = 1.0,
        n::Int64 = 1
    )
        new(
            FoundationLayer(impulses, elements, pressure),
            DynamicLayer(n),
            CognitiveLayer(subjectivity, purpose, time),
            n
        )
    end
end

"""
Compute Intelligence_n = E_n · (1 + F_n) · X · Y · Z · (A · B · C)
"""
function compute_intelligence(axiom::Axiom)::Float64
    foundation_value = compute(axiom.foundation)
    dynamic_value = compute(axiom.dynamic)
    cognitive_value = compute(axiom.cognitive)

    dynamic_value * cognitive_value * foundation_value
end

"""
Evolve the system forward in time
"""
function evolve!(axiom::Axiom, delta_time::Float64 = 1.0)::Float64
    axiom.n += 1
    axiom.dynamic.n = axiom.n
    axiom.cognitive.time += delta_time

    compute_intelligence(axiom)
end

"""
Apply pressure change (e.g., from contradictions or constraints)
"""
function apply_pressure!(axiom::Axiom, pressure_delta::Float64)::Float64
    axiom.foundation.pressure += pressure_delta
    # Ensure pressure stays positive
    axiom.foundation.pressure = max(0.01, axiom.foundation.pressure)

    compute_intelligence(axiom)
end

"""
Adjust subjectivity level (moving toward objectivity or away)
"""
function adjust_subjectivity!(axiom::Axiom, subjectivity_delta::Float64)::Float64
    axiom.cognitive.subjectivity += subjectivity_delta
    # Clamp between 0 and 1
    axiom.cognitive.subjectivity = clamp(axiom.cognitive.subjectivity, 0.0, 1.0)

    compute_intelligence(axiom)
end

"""
Strengthen or weaken purpose alignment
"""
function strengthen_purpose!(axiom::Axiom, purpose_multiplier::Float64)::Float64
    axiom.cognitive.purpose *= purpose_multiplier
    axiom.cognitive.purpose = max(0.01, axiom.cognitive.purpose)

    compute_intelligence(axiom)
end

"""
Get current state of all variables
"""
function get_state(axiom::Axiom)::AxiomState
    foundation_product = compute(axiom.foundation)
    dynamic_product = compute(axiom.dynamic)
    cognitive_product = compute(axiom.cognitive)

    AxiomState(
        axiom.n,
        (
            A_impulses = axiom.foundation.impulses,
            B_elements = axiom.foundation.elements,
            C_pressure = axiom.foundation.pressure,
            product = foundation_product
        ),
        (
            E_n = exponential_growth(axiom.dynamic),
            F_n = fibonacci(axiom.dynamic),
            product = dynamic_product
        ),
        (
            X_subjectivity = axiom.cognitive.subjectivity,
            X_objectivity = objectivity(axiom.cognitive),
            Y_purpose = axiom.cognitive.purpose,
            Z_time = axiom.cognitive.time,
            product = cognitive_product
        ),
        compute_intelligence(axiom)
    )
end

"""
Convert AxiomState to a Dict for JSON serialization
"""
function state_to_dict(state::AxiomState)::Dict{String,Any}
    Dict(
        "n" => state.n,
        "foundation" => Dict(
            "A_impulses" => state.foundation.A_impulses,
            "B_elements" => state.foundation.B_elements,
            "C_pressure" => state.foundation.C_pressure,
            "product" => state.foundation.product
        ),
        "dynamic" => Dict(
            "E_n" => state.dynamic.E_n,
            "F_n" => state.dynamic.F_n,
            "product" => state.dynamic.product
        ),
        "cognitive" => Dict(
            "X_subjectivity" => state.cognitive.X_subjectivity,
            "X_objectivity" => state.cognitive.X_objectivity,
            "Y_purpose" => state.cognitive.Y_purpose,
            "Z_time" => state.cognitive.Z_time,
            "product" => state.cognitive.product
        ),
        "intelligence" => state.intelligence
    )
end

"""
Simulator for running Universal Axiom scenarios
"""
mutable struct AxiomSimulator
    axiom::Axiom
    history::Vector{AxiomState}

    AxiomSimulator(axiom::Axiom) = new(axiom, Vector{AxiomState}())
end

function record_state!(simulator::AxiomSimulator)
    push!(simulator.history, get_state(simulator.axiom))
end

"""
Simulate evolution over multiple time steps
"""
function simulate_evolution(simulator::AxiomSimulator, steps::Int64 = 10, delta_time::Float64 = 1.0)::Vector{AxiomState}
    empty!(simulator.history)
    record_state!(simulator)

    for _ in 1:steps
        evolve!(simulator.axiom, delta_time)
        record_state!(simulator)
    end

    simulator.history
end

"""
Simulate how the system handles contradiction
"""
function simulate_contradiction_resolution(
    simulator::AxiomSimulator,
    initial_pressure::Float64 = 2.0,
    resolution_steps::Int64 = 5
)::Vector{AxiomState}
    empty!(simulator.history)
    record_state!(simulator)

    # Apply initial pressure spike
    apply_pressure!(simulator.axiom, initial_pressure)
    record_state!(simulator)

    # Gradually resolve through objectivity adjustment and pressure release
    for _ in 1:resolution_steps
        # Reduce subjectivity (increase objectivity)
        adjust_subjectivity!(simulator.axiom, -0.1)

        # Release pressure as understanding increases
        pressure_release = -initial_pressure / resolution_steps
        apply_pressure!(simulator.axiom, pressure_release)

        # Evolve forward
        evolve!(simulator.axiom)
        record_state!(simulator)
    end

    simulator.history
end

"""
Calculate coherence metric based on balance of components
"""
function get_coherence_metric(simulator::AxiomSimulator)::Float64
    state = get_state(simulator.axiom)

    objectivity_score = state.cognitive.X_objectivity
    purpose_score = min(state.cognitive.Y_purpose / 2.0, 1.0)
    pressure_score = 1.0 / (1.0 + abs(state.foundation.C_pressure - 1.0))

    (objectivity_score + purpose_score + pressure_score) / 3.0
end

"""
Generate Fibonacci sequence up to n terms
"""
function fibonacci_sequence(n::Int64)::Vector{Int64}
    if n <= 0
        return Int64[]
    elseif n == 1
        return [1]
    end

    sequence = [1, 1]
    for i in 3:n
        push!(sequence, sequence[i-1] + sequence[i-2])
    end

    sequence
end

"""
Proof step aligned with the Universal Axiom.
"""
struct ProofStep
    statement::String
    axiom_insight::String
end

"""
Container for an Erdos problem and its axiom-aligned proof steps.
"""
mutable struct ErdosProblem
    identifier::String
    title::String
    statement::String
    status::String
    axiom_insight::String
    proof_steps::Vector{ProofStep}
end

function ErdosProblem(
    identifier::String,
    title::String,
    statement::String,
    status::String,
    axiom_insight::String
)::ErdosProblem
    ErdosProblem(identifier, title, statement, status, axiom_insight, ProofStep[])
end

function add_proof_step!(problem::ErdosProblem, statement::String, axiom_insight::String)::Nothing
    push!(problem.proof_steps, ProofStep(statement, axiom_insight))
    return nothing
end

"""
Registry for mathematical problems and axiom-aligned proof steps.
"""
mutable struct MathSolutions
    problems::Dict{String, ErdosProblem}
end

function MathSolutions(problems::Vector{ErdosProblem}=ErdosProblem[])::MathSolutions
    mapping = Dict{String, ErdosProblem}()
    for problem in problems
        mapping[problem.identifier] = problem
    end
    MathSolutions(mapping)
end

function erdos_seed()::MathSolutions
    MathSolutions(seed_erdos_problems())
end

function add_problem!(solutions::MathSolutions, problem::ErdosProblem)::Nothing
    solutions.problems[problem.identifier] = problem
    return nothing
end

function get_problem(solutions::MathSolutions, identifier::String)::ErdosProblem
    if !haskey(solutions.problems, identifier)
        error("Unknown problem identifier: $(identifier)")
    end
    solutions.problems[identifier]
end

function list_problems(solutions::MathSolutions)::Vector{ErdosProblem}
    collect(values(solutions.problems))
end

function add_proof_step!(
    solutions::MathSolutions,
    identifier::String,
    statement::String,
    axiom_insight::String
)::Nothing
    problem = get_problem(solutions, identifier)
    add_proof_step!(problem, statement, axiom_insight)
    return nothing
end

function summaries(solutions::MathSolutions)::Vector{NamedTuple}
    [
        (identifier=problem.identifier, title=problem.title, status=problem.status)
        for problem in values(solutions.problems)
    ]
end

function seed_erdos_problems()::Vector{ErdosProblem}
    erdos_straus = ErdosProblem(
        "erdos-straus",
        "Erdos–Straus Conjecture",
        "For every integer n ≥ 2, the rational 4/n can be expressed as the sum of three unit fractions: 4/n = 1/x + 1/y + 1/z for integers x, y, z.",
        "open",
        "The axiom highlights how constraints (C) and growth (E_n, F_n) interact, suggesting structured pathways to decompositions."
    )
    add_proof_step!(
        erdos_straus,
        "Clear denominators to obtain 4xyz = n(xy + xz + yz), exposing the shared ABC constraint.",
        "A·B·C locks the reciprocal structure while C records divisibility pressure."
    )
    add_proof_step!(
        erdos_straus,
        "Partition n into congruence classes to target families where n divides xy + xz + yz.",
        "F_n periodicity mirrors modular cycles, guiding repeatable constructions."
    )
    add_proof_step!(
        erdos_straus,
        "Introduce parameterized families for (x, y, z) that satisfy the cleared equation.",
        "E_n growth supplies expansion room; X and Y keep selections coherent."
    )
    add_proof_step!(
        erdos_straus,
        "Balance denominator growth so x, y, z remain positive and ordered, avoiding runaway residues.",
        "E_n expands search while F_n regulates magnitude."
    )
    add_proof_step!(
        erdos_straus,
        "Cover dense residue families and reduce remaining cases to bounded verification windows.",
        "Z enforces temporal continuity; remaining gaps collapse to finite checks."
    )

    erdos_distinct = ErdosProblem(
        "erdos-distinct-distances",
        "Erdos Distinct Distances Problem",
        "Determine the minimum number of distinct distances defined by n points in the plane.",
        "solved",
        "Balancing combinatorial growth (E_n) with structural regulation (F_n) mirrors the tension between point density and distance diversity."
    )
    add_proof_step!(
        erdos_distinct,
        "Normalize the configuration by translation and scaling to fix baseline spacing.",
        "X, Y, Z align perspective and time scale before counting."
    )
    add_proof_step!(
        erdos_distinct,
        "Count point pairs to relate total pairs to distance multiplicities.",
        "E_n captures pair growth while F_n regulates clustering."
    )
    add_proof_step!(
        erdos_distinct,
        "Apply incidence bounds to limit how often a distance can repeat.",
        "C pressure caps over-concentration in any single distance."
    )
    add_proof_step!(
        erdos_distinct,
        "Construct near-lattice configurations to achieve the lower-bound regime.",
        "A·B·C balances structure so growth matches regulation."
    )
    add_proof_step!(
        erdos_distinct,
        "Conclude the asymptotic bound by matching upper and lower envelopes.",
        "Dynamic layer (E_n, F_n) closes the gap between expansion and constraint."
    )

    erdos_moser = ErdosProblem(
        "erdos-moser",
        "Erdos–Moser Equation",
        "Solve 1^k + 2^k + ... + (m−1)^k = m^k for integers m, k > 1.",
        "partial",
        "E_n scaling intensifies quickly; the axiom suggests using Z to control temporal accumulation and detect singularities."
    )
    add_proof_step!(
        erdos_moser,
        "Compare the power sum to integral bounds to bracket growth of Σ i^k versus m^k.",
        "E_n sets exponential growth while Z tracks accumulation."
    )
    add_proof_step!(
        erdos_moser,
        "Use modular restrictions on k and m to eliminate incompatible residues.",
        "C enforces arithmetic pressure, pruning impossible classes."
    )
    add_proof_step!(
        erdos_moser,
        "Isolate the dominant term by normalizing with m^k and bounding the remainder.",
        "A·B·C stabilizes the foundation as X reduces variance."
    )
    add_proof_step!(
        erdos_moser,
        "Show candidate solutions require extremely tight balance between consecutive powers.",
        "F_n smooths oscillations, exposing near-cancellation requirements."
    )
    add_proof_step!(
        erdos_moser,
        "Reduce remaining candidates to finite computational windows for verification.",
        "Z keeps the search temporal and bounded; Y focuses viable regimes."
    )

    [erdos_straus, erdos_distinct, erdos_moser]
end

# String representation
Base.show(io::IO, axiom::Axiom) = print(io, "Axiom(n=$(axiom.n), Intelligence=$(round(compute_intelligence(axiom), digits=4)))")

end # module

File: src/javascript/universal-axiom.ts
/**
 * The Universal Axiom - Core Implementation (TypeScript)
 *
 * Intelligence_n = E_n · (1 + F_n) · X · Y · Z · (A · B · C)
 *
 * Foundation Layer: Impulses (A), Elements (B), Pressure (C)
 * Dynamic Layer: Exponential Growth (E_n), Fibonacci Sequence (F_n)
 * Cognitive Layer: Subjectivity Scale (X), Why Axis (Y), TimeSphere (Z)
 */

export interface FoundationState {
  impulses: number;   // A - Fundamental drives
  elements: number;   // B - Core components
  pressure: number;   // C - Constraints and forces
}

export interface DynamicState {
  n: number;                    // Current iteration/step
  baseExponential: number;      // Base for exponential growth
}

export interface CognitiveState {
  subjectivity: number;  // X - Subjectivity level (0 = objective, 1 = subjective)
  purpose: number;       // Y - Purpose-driven reasoning strength
  time: number;          // Z - Temporal continuity
}

export interface AxiomState {
  n: number;
  foundation: {
    A_impulses: number;
    B_elements: number;
    C_pressure: number;
    product: number;
  };
  dynamic: {
    E_n: number;
    F_n: number;
    product: number;
  };
  cognitive: {
    X_subjectivity: number;
    X_objectivity: number;
    Y_purpose: number;
    Z_time: number;
    product: number;
  };
  intelligence: number;
}

/**
 * Foundation Layer: A · B · C
 */
export class FoundationLayer {
  constructor(
    public impulses: number,
    public elements: number,
    public pressure: number
  ) {}

  compute(): number {
    return this.impulses * this.elements * this.pressure;
  }
}

/**
 * Dynamic Layer: E_n · (1 + F_n)
 */
export class DynamicLayer {
  constructor(
    public n: number,
    public baseExponential: number = 3
  ) {}

  exponentialGrowth(): number {
    return (2 * Math.pow(this.baseExponential, this.n)) - 1;
  }

  fibonacci(): number {
    if (this.n <= 1) return 1;

    let a = 1;
    let b = 1;

    for (let i = 2; i <= this.n; i++) {
      [a, b] = [b, a + b];
    }

    return b;
  }

  compute(): number {
    const E_n = this.exponentialGrowth();
    const F_n = this.fibonacci();
    return E_n * (1 + F_n);
  }
}

/**
 * Cognitive Layer: X · Y · Z
 */
export class CognitiveLayer {
  constructor(
    public subjectivity: number,  // 0-1 scale
    public purpose: number,
    public time: number
  ) {}

  compute(): number {
    // X represents objectivity: (1 - subjectivity)
    const objectivity = 1 - this.subjectivity;
    return objectivity * this.purpose * this.time;
  }
}

/**
 * The Universal Axiom - Complete Intelligence Model
 *
 * Intelligence_n = E_n · (1 + F_n) · X · Y · Z · (A · B · C)
 */
export class UniversalAxiom {
  foundation: FoundationLayer;
  dynamic: DynamicLayer;
  cognitive: CognitiveLayer;
  n: number;

  constructor({
    impulses = 1.0,
    elements = 1.0,
    pressure = 1.0,
    subjectivity = 0.0,
    purpose = 1.0,
    time = 1.0,
    n = 1
  }: Partial<FoundationState & CognitiveState & { n: number }> = {}) {
    this.foundation = new FoundationLayer(impulses, elements, pressure);
    this.dynamic = new DynamicLayer(n);
    this.cognitive = new CognitiveLayer(subjectivity, purpose, time);
    this.n = n;
  }

  /**
   * Compute Intelligence_n = E_n · (1 + F_n) · X · Y · Z · (A · B · C)
   */
  computeIntelligence(): number {
    const foundationValue = this.foundation.compute();
    const dynamicValue = this.dynamic.compute();
    const cognitiveValue = this.cognitive.compute();

    return dynamicValue * cognitiveValue * foundationValue;
  }

  /**
   * Evolve the system forward in time
   */
  evolve(deltaTime: number = 1.0): number {
    this.n += 1;
    this.dynamic.n = this.n;
    this.cognitive.time += deltaTime;

    return this.computeIntelligence();
  }

  /**
   * Apply pressure change (e.g., from contradictions or constraints)
   */
  applyPressure(pressureDelta: number): number {
    this.foundation.pressure += pressureDelta;
    // Ensure pressure stays positive
    this.foundation.pressure = Math.max(0.01, this.foundation.pressure);

    return this.computeIntelligence();
  }

  /**
   * Adjust subjectivity level (moving toward objectivity or away)
   */
  adjustSubjectivity(subjectivityDelta: number): number {
    this.cognitive.subjectivity += subjectivityDelta;
    // Clamp between 0 and 1
    this.cognitive.subjectivity = Math.max(0.0, Math.min(1.0, this.cognitive.subjectivity));

    return this.computeIntelligence();
  }

  /**
   * Strengthen or weaken purpose alignment
   */
  strengthenPurpose(purposeMultiplier: number): number {
    this.cognitive.purpose *= purposeMultiplier;
    this.cognitive.purpose = Math.max(0.01, this.cognitive.purpose);

    return this.computeIntelligence();
  }

  /**
   * Get current state of all variables
   */
  getState(): AxiomState {
    return {
      n: this.n,
      foundation: {
        A_impulses: this.foundation.impulses,
        B_elements: this.foundation.elements,
        C_pressure: this.foundation.pressure,
        product: this.foundation.compute()
      },
      dynamic: {
        E_n: this.dynamic.exponentialGrowth(),
        F_n: this.dynamic.fibonacci(),
        product: this.dynamic.compute()
      },
      cognitive: {
        X_subjectivity: this.cognitive.subjectivity,
        X_objectivity: 1 - this.cognitive.subjectivity,
        Y_purpose: this.cognitive.purpose,
        Z_time: this.cognitive.time,
        product: this.cognitive.compute()
      },
      intelligence: this.computeIntelligence()
    };
  }

  toString(): string {
    const state = this.getState();
    return `UniversalAxiom(n=${this.n}, Intelligence=${state.intelligence.toFixed(4)})`;
  }
}

/**
 * Simulator for running Universal Axiom scenarios
 */
export class AxiomSimulator {
  private axiom: UniversalAxiom;
  private history: AxiomState[] = [];

  constructor(axiom: UniversalAxiom) {
    this.axiom = axiom;
  }

  recordState(): void {
    this.history.push(this.axiom.getState());
  }

  /**
   * Simulate evolution over multiple time steps
   */
  simulateEvolution(steps: number = 10, deltaTime: number = 1.0): AxiomState[] {
    this.history = [];
    this.recordState();

    for (let i = 0; i < steps; i++) {
      this.axiom.evolve(deltaTime);
      this.recordState();
    }

    return this.history;
  }

  /**
   * Simulate how the system handles contradiction
   */
  simulateContradictionResolution(
    initialPressure: number = 2.0,
    resolutionSteps: number = 5
  ): AxiomState[] {
    this.history = [];
    this.recordState();

    // Apply initial pressure spike
    this.axiom.applyPressure(initialPressure);
    this.recordState();

    // Gradually resolve through objectivity adjustment and pressure release
    for (let i = 0; i < resolutionSteps; i++) {
      // Reduce subjectivity (increase objectivity)
      this.axiom.adjustSubjectivity(-0.1);

      // Release pressure as understanding increases
      const pressureRelease = -initialPressure / resolutionSteps;
      this.axiom.applyPressure(pressureRelease);

      // Evolve forward
      this.axiom.evolve();
      this.recordState();
    }

    return this.history;
  }

  /**
   * Calculate coherence metric based on balance of components
   */
  getCoherenceMetric(): number {
    const state = this.axiom.getState();

    // Coherence is high when:
    // - Objectivity is high (low subjectivity)
    // - Purpose is strong
    // - Pressure is moderate (not too high or low)

    const objectivityScore = state.cognitive.X_objectivity;
    const purposeScore = Math.min(state.cognitive.Y_purpose / 2.0, 1.0);
    const pressureScore = 1.0 / (1.0 + Math.abs(state.foundation.C_pressure - 1.0));

    return (objectivityScore + purposeScore + pressureScore) / 3.0;
  }

  getHistory(): AxiomState[] {
    return this.history;
  }
}

/**
 * Generate Fibonacci sequence up to n terms
 */
export function fibonacciSequence(n: number): number[] {
  if (n <= 0) return [];
  if (n === 1) return [1];

  const sequence = [1, 1];
  for (let i = 2; i < n; i++) {
    sequence.push(sequence[i - 1] + sequence[i - 2]);
  }

  return sequence;
}

export type ProblemStatus = 'open' | 'solved' | 'partial';

export class ProofStep {
  constructor(
    public statement: string,
    public axiomInsight: string
  ) {}

  toJSON(): { statement: string; axiom_insight: string } {
    return {
      statement: this.statement,
      axiom_insight: this.axiomInsight
    };
  }
}

export class ErdosProblem {
  proofSteps: ProofStep[];

  constructor(
    public identifier: string,
    public title: string,
    public statement: string,
    public status: ProblemStatus,
    public axiomInsight: string,
    proofSteps: ProofStep[] = []
  ) {
    this.proofSteps = proofSteps;
  }

  addProofStep(statement: string, axiomInsight: string): void {
    this.proofSteps.push(new ProofStep(statement, axiomInsight));
  }

  addProofSteps(steps: ProofStep[]): void {
    this.proofSteps.push(...steps);
  }

  toJSON(): {
    identifier: string;
    title: string;
    statement: string;
    status: ProblemStatus;
    axiom_insight: string;
    proof_steps: { statement: string; axiom_insight: string }[];
  } {
    return {
      identifier: this.identifier,
      title: this.title,
      statement: this.statement,
      status: this.status,
      axiom_insight: this.axiomInsight,
      proof_steps: this.proofSteps.map((step) => step.toJSON())
    };
  }
}

export class MathSolutions {
  private problems: Map<string, ErdosProblem>;

  constructor(problems: ErdosProblem[] = []) {
    this.problems = new Map();
    problems.forEach((problem) => this.addProblem(problem));
  }

  static erdosSeed(): MathSolutions {
    return new MathSolutions(seedErdosProblems());
  }

  addProblem(problem: ErdosProblem): void {
    this.problems.set(problem.identifier, problem);
  }

  getProblem(identifier: string): ErdosProblem {
    const problem = this.problems.get(identifier);
    if (!problem) {
      throw new Error(`Unknown problem identifier: ${identifier}`);
    }
    return problem;
  }

  listProblems(): ErdosProblem[] {
    return Array.from(this.problems.values());
  }

  addProofStep(identifier: string, statement: string, axiomInsight: string): void {
    const problem = this.getProblem(identifier);
    problem.addProofStep(statement, axiomInsight);
  }

  summaries(): { identifier: string; title: string; status: ProblemStatus }[] {
    return Array.from(this.problems.values()).map((problem) => ({
      identifier: problem.identifier,
      title: problem.title,
      status: problem.status
    }));
  }
}

function seedErdosProblems(): ErdosProblem[] {
  return [
    new ErdosProblem(
      'erdos-straus',
      'Erdos–Straus Conjecture',
      'For every integer n ≥ 2, the rational 4/n can be expressed as the sum of three unit fractions: 4/n = 1/x + 1/y + 1/z for integers x, y, z.',
      'open',
      'The axiom highlights how constraints (C) and growth (E_n, F_n) interact, suggesting structured pathways to decompositions.'
    ),
    new ErdosProblem(
      'erdos-distinct-distances',
      'Erdos Distinct Distances Problem',
      'Determine the minimum number of distinct distances defined by n points in the plane.',
      'solved',
      'Balancing combinatorial growth (E_n) with structural regulation (F_n) mirrors the tension between point density and distance diversity.'
    )
  ];
}
